src/io/flutter/server/vmService/DartVmServiceDebugProcess.java

```diff
@@ -609,4 +533,241 @@ private static String threeSlashize(@NotNull final String uri) {
+                                  VmOpenSourceLocationListener vmOpenSourceLocationListener) {
+    final DartVmServiceListener vmServiceListener =
+      new DartVmServiceListener(this, (DartVmServiceBreakpointHandler)myBreakpointHandlers[0]);
+    final DartVmServiceBreakpointHandler breakpointHandler = (DartVmServiceBreakpointHandler)myBreakpointHandlers[0];
+
+    myVmOpenSourceLocationListener = vmOpenSourceLocationListener;
+    myVmServiceWrapper = new VmServiceWrapper(this, vmService, vmServiceListener, myIsolatesInfo, breakpointHandler);
+
+    final ScriptProvider provider =
+      (isolateId, scriptId) -> myVmServiceWrapper.getScriptSync(isolateId, scriptId);
+
+    mapper.onConnect(provider, myConnector.getRemoteBaseUrl());
+
+    // We disable the remote debug flag so that handleDebuggerConnected() does not echo the stdout and
+    // stderr streams (this would duplicate what we get over daemon logging).
+    remoteDebug = false;
+
+    final FlutterLaunchMode launchMode = FlutterLaunchMode.fromEnv(executionEnvironment);
+    if (launchMode.supportsDebugConnection()) {
+      myVmServiceWrapper.handleDebuggerConnected();
+
+      // TODO(jacobr): the following code is a workaround for issues
+      // auto-resuming isolates paused at their creation while running in
+      // debug mode.
+      // The ideal fix would by to fix VMServiceWrapper so that it checks
+      // for already running isolates like we do here or to refactor where we
+      // create our VmServiceWrapper so we can listen for isolate creation soon
+      // enough that we never miss an isolate creation message.
+      vmService.getVM(new VMConsumer() {
+        @Override
+        public void received(VM vm) {
+          final ElementList<IsolateRef> isolates = vm.getIsolates();
+          // There is a risk the isolate we care about loaded before the call
+          // to handleDebuggerConnected was made and so
+          for (IsolateRef isolateRef : isolates) {
+            vmService.getIsolate(isolateRef.getId(), new VmServiceConsumers.GetIsolateConsumerWrapper() {
+              public void received(Isolate isolate) {
+                final Event event = isolate.getPauseEvent();
+                final EventKind eventKind = event.getKind();
+                if (eventKind == EventKind.PauseStart) {
+                  ApplicationManager.getApplication().invokeLater(() -> {
+                    // We are assuming it is safe to call handleIsolate multiple times.
+                    myVmServiceWrapper.handleIsolate(isolateRef, true);
+                  });
+                }
+              }
+            });
+          }
+        }
+
+        @Override
+        public void onError(RPCError error) {
+          LOG.error(error.toString());
+        }
+      });
+    }
+
+    // We re-enable the remote debug flag so that the service wrapper will call our guessRemoteProjectRoot()
+    // method with the list of loaded libraries for the isolate.
+    remoteDebug = true;
```

Comment:
I'm confused by myRemoteDebug vs remoteDebug. The accessor that (I think) is used by the wrapper tests myRemoteDebug, which is always false. If I'm right, guessRemoteProjectRoot() won't be called. OTOH things seem to work...

Reply:
The remoteDebug logic concerns me and my preference is to remove it.
Essentially currently we are starting with it off to avoid undesirable code paths and then faking that it is off to avoid undesirable code paths.
Now that VmServiceWrapper is in the same project we should be able to clean this up. 

Unless we wanted to support server apps running locally I don't think we need to ever really support anything other than remoteDebug mode.

Reply:
It is worse than that. It starts off but then gets turned on (surprise!) to make sure we went through the right path in code that was formerly in a separate plugin so couldn't be rolled consistently. I've stripped the remote debug flag as I think it just causes confusion and in practice all debugging we want to do for flutter is essentially remote.